### PR TITLE
core, eth/downloader: commit block data using batches

### PR DESCRIPTION
**Pull request by fjl**
In this PR, fast sync state and receipt data is committed in batches of ~100kB. This size was determined using experiments with https://github.com/fjl/goleveldb-bench. Another related optimization in this PR removes random reads in HasBlock and HasHeader.

The PR also restores batch writes of state data for full imports and writes the block, its receipts and most other data in the same batch.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -693,108 +696,70 @@ func SetReceiptsData(config *params.ChainConfig, block *types.Block, receipts ty
...
-
-	errs, failed := make([]error, len(tasks)), int32(0)
-	process := func(worker int) {
-		for index := range tasks {
-			block, receipts := blockChain[index], receiptChain[index]
+	batch := bc.chainDb.NewBatch()
+	for i := 0; i < len(blockChain); i++ {
+		block, receipts := blockChain[i], receiptChain[i]
+		if i > 0 && (block.NumberU64() != blockChain[i-1].NumberU64()+1 || block.ParentHash() != blockChain[i-1].Hash()) {
+			// Chain broke ancestry, log a messge (programming error) and skip insertion
```
Comment (holiman) [inline core/blockchain.go]: s/messge/message

File: core/blockchain.go [hunk 2]
```diff
@@ -693,108 +696,70 @@ func SetReceiptsData(config *params.ChainConfig, block *types.Block, receipts ty
...
-		tasks <- i
-	}
-	close(tasks)
-
-	errs, failed := make([]error, len(tasks)), int32(0)
-	process := func(worker int) {
-		for index := range tasks {
-			block, receipts := blockChain[index], receiptChain[index]
+	batch := bc.chainDb.NewBatch()
+	for i := 0; i < len(blockChain); i++ {
```
Comment (holiman) [inline core/blockchain.go]: Why did you move this pre-checker for inside the main loop?

File: core/blockchain.go [hunk 3]
```diff
@@ -693,108 +696,70 @@ func SetReceiptsData(config *params.ChainConfig, block *types.Block, receipts ty
...
+		block, receipts := blockChain[i], receiptChain[i]
+		if i > 0 && (block.NumberU64() != blockChain[i-1].NumberU64()+1 || block.ParentHash() != blockChain[i-1].Hash()) {
+			// Chain broke ancestry, log a messge (programming error) and skip insertion
+			log.Error("Non contiguous receipt insert", "number", block.Number(), "hash", block.Hash(), "parent", block.ParentHash(),
+				"prevnumber", blockChain[i-1].Number(), "prevhash", blockChain[i-1].Hash())
+			return 0, fmt.Errorf("non contiguous insert: item %d is #%d [%x…], item %d is #%d [%x…] (parent [%x…])", i-1, blockChain[i-1].NumberU64(),
```
Comment (holiman) [inline core/blockchain.go]: What's the consequence if it exits here on block X, after having already written a couple of blocks to `batch` ?

Comment (fjl) [reply core/blockchain.go]: Blocks will be written, but not marked active. This should be fine, but I've moved the check back anyway.

**Comment by fjl** [comment]
@holiman PTAL

**Comment by fjl** [comment]
It's really hard to measure the impact of this change. Package core benchmarks show a slight improvement for disk databases.

```
name                                old time/op    new time/op    delta
InsertChain_empty_memdb-8             95.5µs ± 1%    96.8µs ± 1%     ~     (p=0.095 n=5+5)
InsertChain_empty_diskdb-8             265µs ± 5%     209µs ± 1%  -21.02%  (p=0.008 n=5+5)
InsertChain_valueTx_memdb-8            185µs ± 2%     180µs ± 1%   -2.69%  (p=0.008 n=5+5)
InsertChain_valueTx_diskdb-8           379µs ± 1%     296µs ± 2%  -21.90%  (p=0.008 n=5+5)
InsertChain_valueTx_100kB_memdb-8      980µs ± 1%     998µs ± 4%     ~     (p=0.111 n=4+5)
InsertChain_valueTx_100kB_diskdb-8    1.42ms ± 1%    1.34ms ± 4%   -5.84%  (p=0.008 n=5+5)
InsertChain_uncles_memdb-8             195µs ± 2%     196µs ± 1%     ~     (p=0.690 n=5+5)
InsertChain_uncles_diskdb-8            360µs ± 1%     306µs ± 1%  -14.82%  (p=0.008 n=5+5)
InsertChain_ring200_memdb-8           84.9ms ± 2%    84.6ms ± 1%     ~     (p=1.000 n=5+5)
InsertChain_ring200_diskdb-8          83.2ms ± 1%    78.6ms ± 1%   -5.52%  (p=0.008 n=5+5)
InsertChain_ring1000_memdb-8          89.9ms ± 3%    88.7ms ± 1%     ~     (p=0.310 n=5+5)
InsertChain_ring1000_diskdb-8         89.0ms ± 1%    82.9ms ± 1%   -6.81%  (p=0.008 n=5+5)
```

File: ethdb/memory_database.go [hunk 4]
```diff
@@ -45,6 +45,14 @@ func (db *MemDatabase) Put(key []byte, value []byte) error {
 	return nil
 }
 
+func (db *MemDatabase) Has(key []byte) (bool, error) {
+	db.lock.Lock()
+	defer db.lock.Unlock()
```
Comment (karalabe) [inline ethdb/memory_database.go]: Read locking is enough.

File: ethdb/interface.go [hunk 5]
```diff
@@ -16,15 +16,29 @@
...
 	Close()
 	NewBatch() Batch
 }
 
+// Batch is a write-only database that commits changes to its host database
+// when Write is called. Batch cannot be used concurrently.
 type Batch interface {
-	Put(key, value []byte) error
+	Putter
+	ValueSize() int // amount of data in the batch
```
Comment (karalabe) [inline ethdb/interface.go]: Can't we just call this `Size`?

File: eth/downloader/statesync.go [hunk 6]
```diff
@@ -187,10 +188,13 @@ func (d *Downloader) runStateSync(s *stateSync) *stateSync {
 type stateSync struct {
 	d *Downloader // Downloader instance to access and manage current peerset
 
-	sched  *state.StateSync           // State trie sync scheduler defining the tasks
+	sched  *trie.TrieSync             // State trie sync scheduler defining the tasks
 	keccak hash.Hash                  // Keccak256 hasher to verify deliveries with
 	tasks  map[common.Hash]*stateTask // Set of tasks currently queued for retrieval
 
+	numUncommitted   int
+	bytesUncommitted int
```
Comment (karalabe) [inline eth/downloader/statesync.go]: I'd suggest renaming to `uncommittedNodes` and `uncommittedBytes`. Looks and aligns better. Names form a "group" having the same prefix.

File: core/blockchain.go [hunk 7]
```diff
@@ -839,9 +809,17 @@ func (bc *BlockChain) WriteBlock(block *types.Block) (status WriteStatus, err er
 	if err := bc.hc.WriteTd(block.Hash(), block.NumberU64(), externTd); err != nil {
 		log.Crit("Failed to write block total difficulty", "err", err)
 	}
-	if err := WriteBlock(bc.chainDb, block); err != nil {
+	// Write other block data using a batch.
+	batch := bc.chainDb.NewBatch()
+	if err := WriteBlock(batch, block); err != nil {
 		log.Crit("Failed to write block contents", "err", err)
```
Comment (karalabe) [inline core/blockchain.go]: Perhaps `return NonStatTy, err` here too?

File: core/blockchain.go [hunk 8]
```diff
@@ -517,8 +516,12 @@ func (bc *BlockChain) GetBodyRLP(hash common.Hash) rlp.RawValue {
...
 // HasBlock checks if a block is fully present in the database or not, caching
 // it if present.
-func (bc *BlockChain) HasBlock(hash common.Hash) bool {
-	return bc.GetBlockByHash(hash) != nil
+func (bc *BlockChain) HasBlock(hash common.Hash, number uint64) bool {
+	if bc.blockCache.Contains(hash) {
+		return true
+	}
+	ok, _ := bc.chainDb.Has(blockBodyKey(hash, number))
+	return ok
```
Comment (karalabe) [inline core/blockchain.go]: This new code doesn't cache the response any more. Should we? If not, please fix the comment. If yes, fix the code.

File: core/headerchain.go [hunk 9]
```diff
@@ -361,10 +361,13 @@ func (hc *HeaderChain) GetHeaderByHash(hash common.Hash) *types.Header {
...
-// it if present.
-func (hc *HeaderChain) HasHeader(hash common.Hash) bool {
-	return hc.GetHeaderByHash(hash) != nil
+// HasHeader checks if a block header is present in the database or not.
+func (hc *HeaderChain) HasHeader(hash common.Hash, number uint64) bool {
+	if hc.numberCache.Contains(hash) || hc.headerCache.Contains(hash) {
+		return true
+	}
+	ok, _ := hc.chainDb.Has(headerKey(hash, number))
+	return ok
```
Comment (karalabe) [inline core/headerchain.go]: This new code doesn't cache the response any more. Should we? If not, please fix the comment. If yes, fix the code.

File: eth/downloader/statesync.go (hunk 6 above)
Comment (fjl) [reply eth/downloader/statesync.go]: I want to keep these names.

File: ethdb/interface.go (hunk 5 above)
Comment (fjl) [reply ethdb/interface.go]: The size is really just the sum of value sizes. I wanted to avoid confusion with leveldb batch `Len` method, which returns number of items.

**Comment by fjl** [comment]
@karalabe PTAL
